File: kartwheel/app/src/main/java/us/handstand/kartwheel/activity/TicketActivity.java
package us.handstand.kartwheel.activity;


import android.content.Context;
import android.content.Intent;
import android.content.res.ColorStateList;
import android.os.Bundle;
import android.support.annotation.ColorRes;
import android.support.annotation.IntDef;
import android.support.annotation.StringRes;
import android.support.v4.app.Fragment;
import android.support.v4.view.ViewCompat;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.AppCompatButton;
import android.view.View;
import android.widget.TextView;
import android.widget.Toast;

import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;
import us.handstand.kartwheel.R;
import us.handstand.kartwheel.fragment.AlreadyClaimedFragment;
import us.handstand.kartwheel.fragment.CodeEntryFragment;
import us.handstand.kartwheel.fragment.CriticalInformationFragment;
import us.handstand.kartwheel.fragment.TOSFragment;
import us.handstand.kartwheel.fragment.WelcomeFragment;
import us.handstand.kartwheel.layout.ViewUtil;
import us.handstand.kartwheel.model.AndroidStorage;
import us.handstand.kartwheel.model.Race;
import us.handstand.kartwheel.model.Ticket;
import us.handstand.kartwheel.model.User;
import us.handstand.kartwheel.network.API;

public class TicketActivity extends AppCompatActivity implements View.OnClickListener {
    public interface TicketFragment extends View.OnClickListener {
    }

    @IntDef({TOS, CODE_ENTRY, CRITICAL_INFO, WELCOME, ALREADY_CLAIMED, FORFEIT, GAME_INFO})
    private @interface FragmentType {
    }

    private static final String INTENT_EXTRA_FRAGMENT_TYPE = "fragment_type";
    private static final int TOS = 0;
    private static final int CODE_ENTRY = 1;
    private static final int CRITICAL_INFO = 2;
    private static final int WELCOME = 3;
    private static final int ALREADY_CLAIMED = 4;
    private static final int FORFEIT = 5;
    private static final int GAME_INFO = 6;

    @BindView(R.id.title_text) TextView title;
    @BindView(R.id.button) AppCompatButton button;
    @BindView(R.id.additional_button) AppCompatButton additionalButton;
    private TicketFragment ticketFragment;

    static Intent getStartIntent(Context context) {
        Intent intent = new Intent(context, TicketActivity.class);
        intent.putExtra(INTENT_EXTRA_FRAGMENT_TYPE, TOS);
        return intent;
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_ticket);
        ButterKnife.bind(this);
        button.setOnClickListener(this);
        additionalButton = ViewUtil.findView(this, R.id.additional_button);
        additionalButton.setOnClickListener(this);
        showFragment(getCurrentFragmentType());
    }

    public void showFragment(@FragmentType int type) {
        title.setText(R.string.app_name);
        additionalButton.setVisibility(View.GONE);
        switch (type) {
            case TOS:
                ticketFragment = new TOSFragment();
                button.setText(R.string.scroll_down);
                break;
            case CODE_ENTRY:
                ticketFragment = new CodeEntryFragment();
                button.setText(R.string.lets_go);
                break;
            case ALREADY_CLAIMED:
                title.setText(R.string.already_claimed_title);
                button.setText(R.string.try_different_code);
                setAdditionalButtonState(R.color.green, R.string.contact_us, true);
                ticketFragment = new AlreadyClaimedFragment();
                break;
            case CRITICAL_INFO:
                title.setText(R.string.welcome);
                button.setText(R.string.next);
                ticketFragment = new CriticalInformationFragment();
                break;
            case WELCOME:
                title.setText(R.string.welcome);
                ticketFragment = new WelcomeFragment();
                setButtonState(R.color.grey_button_disabled, R.string.im_ready, false);
                break;
        }
        if (ticketFragment != null) {
            getIntent().putExtra(INTENT_EXTRA_FRAGMENT_TYPE, type);
            if (!isFinishing() && !getSupportFragmentManager().isDestroyed()) {
                getSupportFragmentManager().beginTransaction().replace(R.id.fragment, (Fragment) ticketFragment).commit();
            }
        }
    }

    @FragmentType
    private int getCurrentFragmentType() {
        @FragmentType int currentFragmentType = getIntent().getIntExtra(INTENT_EXTRA_FRAGMENT_TYPE, TOS);
        return currentFragmentType;
    }

    public void setButtonState(@ColorRes int color, @StringRes int textRes, boolean enabled) {
        //noinspection unchecked
        ViewCompat.setBackgroundTintList(button, ColorStateList.valueOf(getResources().getColor(color)));
        button.setText(textRes);
        button.setEnabled(enabled);
    }

    public void setAdditionalButtonState(@ColorRes int color, @StringRes int textRes, boolean enabled) {
        //noinspection unchecked
        ViewCompat.setBackgroundTintList(additionalButton, ColorStateList.valueOf(getResources().getColor(color)));
        additionalButton.setText(textRes);
        additionalButton.setEnabled(enabled);
        if (enabled) {
            additionalButton.setVisibility(View.VISIBLE);
        }
    }

    @Override
    public void onClick(View v) {
        if (!v.isEnabled() || !(v instanceof AppCompatButton)) {
            return;
        }

        ticketFragment.onClick(v);
        switch (getCurrentFragmentType()) {
            case TOS:
                showFragment(CODE_ENTRY);
                break;
            case CODE_ENTRY:
                if (getIntent().hasExtra(Ticket.CODE)) {
                    API.claimTicket(getIntent().getStringExtra(Ticket.CODE), new API.APICallback<Ticket>() {
                        @Override
                        public void onSuccess(Ticket ticket) {
                            if (ticket.isClaimed()) {
                                API.getRaces(AndroidStorage.getString(AndroidStorage.EVENT_ID), new API.APICallback<List<Race>>() {
                                    @Override
                                    public void onSuccess(List<Race> response) {
                                        showFragment(GAME_INFO);
                                    }

                                    @Override
                                    public void onFailure(int errorCode, String errorResponse) {
                                        Toast.makeText(TicketActivity.this, "Failed to get races " + errorResponse, Toast.LENGTH_LONG).show();
                                    }
                                });
                            } else {
                                runOnUiThread(() -> showFragment(CRITICAL_INFO));
                            }
                        }

                        @Override
                        public void onFailure(int errorCode, String errorResponse) {
                            if (errorCode == 409) {
                                runOnUiThread(() -> showFragment(ALREADY_CLAIMED));
                            } else if (errorResponse != null) {
                                Toast.makeText(TicketActivity.this, errorResponse, Toast.LENGTH_LONG).show();
                            }
                        }
                    });
                }
                break;
            case ALREADY_CLAIMED:
                if (v.getId() == R.id.additional_button) {
                    final Intent emailIntent = new Intent(Intent.ACTION_SEND)
                            .setType("plain/text")
                            .putExtra(Intent.EXTRA_EMAIL, new String[]{getResources().getString(R.string.support_email)})
                            .putExtra(Intent.EXTRA_SUBJECT, getResources().getString(R.string.contact_us_subject_line))
                            .putExtra(Intent.EXTRA_TEXT, getResources().getString(R.string.contact_us_body, getIntent().getStringExtra(Ticket.CODE)));
                    startActivity(Intent.createChooser(emailIntent, getResources().getString(R.string.contact_us)));
                } else {
                    showFragment(CODE_ENTRY);
                }
                break;
            case CRITICAL_INFO:
                if (getIntent().hasExtra(User.PANCAKEORWAFFLE) &&
                        getIntent().hasExtra(User.CHARMANDERORSQUIRTLE)) {
                    showFragment(WELCOME);
                }
                break;
            case WELCOME:
                if (getIntent().hasExtra(User.BIRTH) &&
                        getIntent().hasExtra(User.CELL) &&
                        getIntent().hasExtra(User.EMAIL) &&
                        getIntent().hasExtra(User.FIRSTNAME) &&
                        getIntent().hasExtra(User.LASTNAME) &&
                        getIntent().hasExtra(User.NICKNAME)) {
                    User user = User.FACTORY.creator.create(AndroidStorage.getString(AndroidStorage.USER_ID),
                            null,  // authToken
                            getIntent().getStringExtra(User.BIRTH),
                            getIntent().getStringExtra(User.CELL),
                            getIntent().getStringExtra(User.CHARMANDERORSQUIRTLE),
                            getIntent().getStringExtra(User.EMAIL),
                            AndroidStorage.getString(AndroidStorage.EVENT_ID),
                            null, // facetime count
                            getIntent().getStringExtra(User.FIRSTNAME),
                            null, // imageUrl
                            getIntent().getStringExtra(User.LASTNAME),
                            null, // miniGameId
                            getIntent().getStringExtra(User.NICKNAME),
                            getIntent().getStringExtra(User.PANCAKEORWAFFLE),
                            null, // device token
                            null, // push enabled
                            null, // race id
                            null, // referral type
                            null, // team id
                            null, // total anti miles
                            null, // total distance miles
                            null // updated at
                    );
                    API.updateUser(user, new API.APICallback<User>() {
                        @Override
                        public void onSuccess(User response) {
                            runOnUiThread(() -> showFragment(GAME_INFO));
                        }

                        @Override
                        public void onFailure(int errorCode, String errorResponse) {
                            Toast.makeText(TicketActivity.this, "Unable to create user: " + errorResponse, Toast.LENGTH_SHORT).show();
                        }
                    });
                }
                break;
        }
    }
}

File: kartwheel/app/src/main/kotlin/us/handstand/kartwheel/activity/TicketActivity.kt
package us.handstand.kartwheel.activity

import android.app.Activity
import android.content.Intent
import android.os.Bundle
import android.support.v4.app.Fragment
import android.support.v7.app.AppCompatActivity
import android.view.View
import android.widget.TextView
import us.handstand.kartwheel.KartWheel
import us.handstand.kartwheel.R
import us.handstand.kartwheel.controller.TicketController
import us.handstand.kartwheel.controller.TicketController.Companion.ALREADY_CLAIMED
import us.handstand.kartwheel.controller.TicketController.Companion.CODE_ENTRY
import us.handstand.kartwheel.controller.TicketController.Companion.CRITICAL_INFO
import us.handstand.kartwheel.controller.TicketController.Companion.ERROR
import us.handstand.kartwheel.controller.TicketController.Companion.FORFEIT
import us.handstand.kartwheel.controller.TicketController.Companion.FragmentType
import us.handstand.kartwheel.controller.TicketController.Companion.GAME_INFO
import us.handstand.kartwheel.controller.TicketController.Companion.NONE
import us.handstand.kartwheel.controller.TicketController.Companion.ONBOARDING
import us.handstand.kartwheel.controller.TicketController.Companion.RACE_LIST
import us.handstand.kartwheel.controller.TicketController.Companion.TOS
import us.handstand.kartwheel.controller.TicketController.Companion.WELCOME
import us.handstand.kartwheel.fragment.ticket.*
import us.handstand.kartwheel.layout.KartButton
import us.handstand.kartwheel.layout.ViewUtil
import us.handstand.kartwheel.layout.setCandyCaneBackground
import us.handstand.kartwheel.model.Database
import us.handstand.kartwheel.model.Storage
import us.handstand.kartwheel.util.SnackbarUtil

class TicketActivity : AppCompatActivity(), View.OnClickListener, TicketController.Companion.TicketStepCompletionListener {

    interface TicketFragment {

        fun getTitleResId(): Int

        fun getAdvanceButtonTextResId(): Int

        fun getAdvanceButtonColor(): Int {
            return R.color.grey_button_disabled
        }

        fun getAdvanceButtonLoadingColor(): Int {
            return android.R.color.black
        }

        fun isAdvanceButtonEnabled(): Boolean {
            return false
        }

        fun canAdvanceToNextStep(): Boolean {
            return true
        }

        fun getActivity(): Activity

        companion object {
            fun getFragment(@FragmentType fragmentType: Long): TicketFragment {
                when (fragmentType) {
                    TOS -> return TOSFragment()
                    CODE_ENTRY -> return CodeEntryFragment()
                    ALREADY_CLAIMED -> return AlreadyClaimedFragment()
                    CRITICAL_INFO -> return CriticalInfoFragment()
                    WELCOME -> return WelcomeFragment()
                    GAME_INFO -> return GameInfoFragment()
                    FORFEIT -> return ForfeitFragment()
                }
                return TOSFragment()
            }
        }

        val ticketController: TicketController
            get() {
                return (getActivity() as TicketActivity).ticketController
            }
    }

    val ticketController = TicketController(Database.get(), this)
    private var ticketFragment: TicketFragment? = null
    internal var title: TextView? = null
    internal var button: KartButton? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_ticket)
        findViewById(R.id.parent).setCandyCaneBackground(android.R.color.white, R.color.textLightGrey_40p)
        title = ViewUtil.findView(this, R.id.title_text)
        button = ViewUtil.findView(this, R.id.button)
        button!!.setOnClickListener(this)
        ticketController.transition(NONE, Storage.lastTicketState)
    }

    override fun onDestroy() {
        ticketController.onDestroy()
        super.onDestroy()
    }

    override fun showDialog(@FragmentType step: Long, message: String) {
        runOnUiThread {
            if (step == CODE_ENTRY && ticketFragment is CodeEntryFragment) {
                button?.loading = false
                (ticketFragment as CodeEntryFragment).setProgressVisibility(View.INVISIBLE)
            }
            SnackbarUtil.show(this, message)
        }
    }

    override fun showNextStep(@FragmentType previous: Long, @FragmentType next: Long) {
        runOnUiThread {
            // Make sure that we're starting with fresh data.
            if (next == CODE_ENTRY) {
                KartWheel.logout(lastTicketState = CODE_ENTRY)
            }

            if (next != ERROR) {
                ViewUtil.hideKeyboard(this)
                button?.loading = false
                Storage.lastTicketState = next
                if (next == RACE_LIST) {
                    startActivity(Intent(this, LoggedInActivity::class.java))
                    finish()
                } else if (next == ONBOARDING) {
                    startActivity(Intent(this, OnboardingActivity::class.java))
                    finish()
                } else if (!isFinishing && !supportFragmentManager.isDestroyed) {
                    ticketFragment = TicketFragment.getFragment(next)
                    title!!.text = resources.getString(ticketFragment!!.getTitleResId())
                    onTicketFragmentStateChanged()
                    supportFragmentManager.beginTransaction().replace(R.id.fragment, ticketFragment as Fragment?).commit()
                }
            }
        }
    }

    override fun onTicketFragmentStateChanged() {
        if (ticketFragment != null) {
            try {
                ViewUtil.setButtonState(resources, button,
                        ticketFragment!!.getAdvanceButtonColor(),
                        ticketFragment!!.getAdvanceButtonLoadingColor(),
                        ticketFragment!!.getAdvanceButtonTextResId(),
                        ticketFragment!!.isAdvanceButtonEnabled()
                )
            } catch (e: UninitializedPropertyAccessException) {
                ViewUtil.setButtonState(resources, button, R.color.grey_button_disabled, android.R.color.black, R.string.next, false)
            }
        }
    }

    override fun onClick(v: View) {
        if (v.isEnabled && v.id == R.id.button && ticketFragment!!.isAdvanceButtonEnabled() && ticketFragment!!.canAdvanceToNextStep()) {
            if (ticketFragment is CodeEntryFragment) {
                button!!.loading = true
                (ticketFragment as CodeEntryFragment).setProgressVisibility(View.VISIBLE)
            }
            val lastState = Storage.lastTicketState
            ticketController.onStepCompleted(lastState)
        }
    }
}
